core/synch.c

```diff
@@ -1827,11 +1826,11 @@ send_all_other_threads_native(void)
 #endif
 
     /* Suspend all threads except those trying to synch with us */
-    DEBUG_DECLARE(ok =)
-        synch_with_all_threads(desired_state, &threads, &num_threads,
+    bool ok = synch_with_all_threads(desired_state, &threads, &num_threads,
                                THREAD_SYNCH_NO_LOCKS_NO_XFER,
                                THREAD_SYNCH_SUSPEND_FAILURE_IGNORE);
-    ASSERT(ok);
+    DR_ASSERT_MSG(ok, "send_all_other_threads_native: "
```

Comment:
DR_ASSERT_MSG is for clients.  Use ASSERT_MESSAGE() for a debug-build assert with a string or just ASSERT(ok && "msg here").  But, I thought this CL wanted to have a release-build check.  Inside DR we would not call that an "assert" as that implies debug-build-only.  Existing release-build exits on fatal errors are mostly limited to external usage errors through external_error.  There are only a handful of internal errors and they have their own events.mc entries and generally call their own SYSLOG, report_dynamorio_problem() for crashes that want a callstack and lib list or os_dump_core call w/ filter check o/w, and os_terminate().  So, I would suggest adding a new events.mc message.  Maybe a helper parallel to external_error could be added that took in the msg id and the dumpcore filter.

Reply:
There exists an internal_error function which calls report_dynamorio_problem (which does some sort of syslog along with an optional core dump), and then does a soft_terminate. It appears to just be defined for DEBUG or INTERNAL builds. Should I just expose that function in all cases so that it can be used here? The syslog message that is printed for internal_error is rather generic (depending on parameters passed to the function) so maybe you'd rather me implement a new function similar to internal_error which took its own events.mc symbol to print out.

Reply:
In DR we historically have not wanted a general release-build sanity check call: the approach has been to explicitly handle anything that might happen from the app, and use debug-build-only (for perf reasons) sanity checks to catch internal logic errors or client API errors.  For the rare cases where we can't handle something and can't continue, such as a crash in our own code, or unrecoverable OOM, we exit with an explicit code and diagnostics.  In keeping with that I would vote for not adding a "release assert" with internal_error (and for changing the commit message title here) and keeping each release build exit with its own code.

Reply:
OK, if I read what you requested correctly, it seems like you wanted me to still create some parallel to external_error/internal_error for this fatal error. So I wrote a fatal_error macro to encapsulate what I think you're asking. If I've misunderstood please let me know.

Also, let me know where the exit codes should be stored. I was hoping that the events.mc would have a way to create arbitrary defines but I couldn't figure out how to do that. So I just defined the exit code for the specific fatal message in utils.h; I'm hoping you have a better idea here.